core/include/gnuradio-4.0/Block.hpp

```diff
@@ -1356,6 +1431,72 @@ struct Block : protected std::tuple<Arguments...> {
             return workInternal(requested_work);
         }
     }
+
+    void
+    processMessages(MsgPortInNamed<"__Builtin"> &port, std::span<const Message> messages) {
+        fmt::print("{} got a message\n", self().unique_name);
+        if (std::addressof(port) != std::addressof(builtinMessagePorts.input)) {
+            fmt::print("{} got a message on a wrong port\n", self().unique_name);
+            return;
+        }
+
+        for (const auto &message : messages) {
+            const auto kind   = messageField<std::string>(message, gr::message::key::Kind).value();
+            const auto target = messageField<std::string>(message, gr::message::key::Target);
+
+            if (target && !target->empty() && *target != self().unique_name) {
+                fmt::print("{} message {} is not for me, but for {}", self().unique_name, kind, *target);
+                continue;
+            }
+
+            if (kind == gr::message::kind::Control) {
+                const auto what = messageField<std::string>(message, gr::message::key::What).value();
+                // fmt::print("{} got a control message {}\n", self().unique_name, what);
+                if constexpr (requires { self().start(); })
+                    if (what == gr::message::control::Start) {
+                        self().start();
+                    }
+                if constexpr (requires { self().stop(); })
+                    if (what == gr::message::control::Stop) {
+                        self().stop();
+                    }
+                if constexpr (requires { self().pause(); })
+                    if (what == gr::message::control::Pause) {
+                        self().pause();
+                    }
+                if constexpr (requires { self().resume(); })
+                    if (what == gr::message::control::Resume) {
+                        self().resume();
+                    }
+                if constexpr (requires { self().reset(); })
+                    if (what == gr::message::control::Reset) {
+                        self().reset();
+                    }
+
```

Comment:
Meant more as an example than a real world use-case.

Reply:
OK. The lifecycle should be directly invoked by the scheduler -> graph -> blocks because these need to be synchronously processed (contract/guarantee) whereas the message ports are by designed asynchronous and their message processed at -- potentially -- undefined moment of time.